library/ziti-socket.c

```diff
@@ -280,38 +338,135 @@ static void on_bridge_close(void *ctx) {
+            break;
+        default:
+            return NULL;
+    }
+
+    int i;
+    MODEL_MAP_FOREACH(service, intercept, &wrap->intercepts) {
+        bool proto_match = false;
+        bool port_match = false;
+        bool host_match = false;
+        for (i = 0; !proto_match && intercept->protocols[i] != NULL; i++) {
+            proto_match = strcasecmp(proto, intercept->protocols[i]) == 0;
+        }
+        if (!proto_match) continue;
+
+        for (i = 0; !port_match && intercept->port_ranges[i] != NULL; i++) {
+            ziti_port_range *range = intercept->port_ranges[i];
+            port_match = range->low <= port && port <= range->high;
+        }
+        if (!port_match) continue;
+
+        host_match = ziti_address_match_array(host, intercept->addresses);
+
+        if (host_match)
+            return service;
+    }
+    return NULL;
+}
+
 static void do_ziti_connect(struct dial_req_s *req, future_t *f, uv_loop_t *l) {
     ziti_sock_t *zs = model_map_get_key(&ziti_sockets, &req->fd, sizeof(req->fd));
     if (zs == NULL) {
-        ZITI_LOG(WARN, "socket %ld not found", req->fd);
+        ZITI_LOG(WARN, "socket %lu not found", (unsigned long)req->fd);
         fail_future(f, -EBADF);
     } else if (zs->f != NULL) {
         fail_future(f, -EALREADY);
     } else {
-        zs->ztx = req->ztx;
         zs->f = f;
-        ziti_conn_init(req->ztx, &zs->conn, zs);
-        ziti_dial(zs->conn, req->service, on_ziti_connect, NULL);
+
+        int proto = 0;
+        socklen_t optlen = sizeof(proto);
+        if (getsockopt(req->fd, SOL_SOCKET, SO_TYPE, &proto, &optlen)) {
+            ZITI_LOG(WARN, "unknown socket type fd[%d]: %d(%s)", req->fd, errno, strerror(errno));
+        }
+
+        if (req->ztx == NULL) {
+            MODEL_MAP_FOR(it, ziti_contexts) {
+                ztx_wrap_t *wrap = model_map_it_value(it);
+                const char *service_name = find_service(wrap, proto, req->host, req->port);
+
+                if (service_name != NULL) {
+                    req->ztx = wrap->ztx;
+                    req->service = service_name;
+                    break;
+                }
+            }
+        }
+
+        if (req->ztx != NULL) {
+            const char *proto_str = proto == SOCK_DGRAM ? "udp" : "tcp";
+
+            ziti_conn_init(req->ztx, &zs->conn, zs);
+            char app_data[1024];
+            size_t len = snprintf(app_data, sizeof(app_data),
+                                  "{\"dst_protocol\": \"%s\", \"dst_hostname\": \"%s\", \"dst_port\": \"%u\"}",
+                                  proto_str, req->host, req->port);
+            ziti_dial_opts opts = {
+                    .app_data = app_data,
+                    .app_data_sz = len,
+            };
+            ZITI_LOG(INFO, "zs[%p]->f[%p]", zs, zs->f);
+            ziti_dial_with_options(zs->conn, req->service, &opts, on_ziti_connect, NULL);
+        } else {
+            fail_future(f, -ECONNREFUSED);
```

Comment:
i think you should log this too. this seems like a strange situation to be in

Reply:
it is not that strange: the address you're connecting to is not in any service intercepts